sagemaker-spark/pyspark_mnist/pyspark_mnist_kmeans.ipynb

```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "# replace this with your role ARN\n",
+    "kmeans_estimator = KMeansSageMakerEstimator(\n",
+    "    sagemakerRole=IAMRole(role),\n",
+    "    trainingInstanceType='ml.p2.xlarge',\n",
+    "    trainingInstanceCount=1,\n",
+    "    endpointInstanceType='ml.c4.xlarge',\n",
+    "    endpointInitialInstanceCount=1)\n",
+    "\n",
+    "kmeans_estimator.setK(10)\n",
+    "kmeans_estimator.setFeatureDim(784)\n",
+    "\n",
+    "# train\n",
+    "model = kmeans_estimator.fit(trainingData)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Inference\n",
+    "\n",
+    "Now we transform our DataFrame.\n",
+    "To do this, we serialize each row's \"features\" Vector of Doubles into a Protobuf format for inference against the Amazon SageMaker Endpoint. We deserialize the Protobuf responses back into our DataFrame:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "transformedData = model.transform(testData)\n",
+    "\n",
+    "transformedData.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from pyspark.sql.types import DoubleType\n",
+    "import matplotlib.pyplot as plt\n",
+    "import numpy as np\n",
+    "\n",
+    "# helper function to display a digit\n",
+    "def show_digit(img, caption='', xlabel='', subplot=None):\n",
+    "    if subplot==None:\n",
+    "        _,(subplot)=plt.subplots(1,1)\n",
+    "    imgr=img.reshape((28,28))\n",
+    "    subplot.axes.get_xaxis().set_ticks([])\n",
+    "    subplot.axes.get_yaxis().set_ticks([])\n",
+    "    plt.title(caption)\n",
+    "    plt.xlabel(xlabel)\n",
+    "    subplot.imshow(imgr, cmap='gray')\n",
+    "\n",
+    "images = np.array(transformedData.select(\"features\").cache().take(250))\n",
+    "clusters = transformedData.select(\"closest_cluster\").cache().take(250)\n",
+    "\n",
+    "for cluster in range(10):\n",
+    "    print('\\n\\n\\nCluster {}:'.format(int(cluster)))\n",
+    "    digits = [ img for l, img in zip(clusters, images) if int(l.closest_cluster) == cluster ]\n",
+    "    height=((len(digits)-1)//5)+1\n",
+    "    width=5\n",
+    "    plt.rcParams[\"figure.figsize\"] = (width,height)\n",
+    "    _, subplots = plt.subplots(height, width)\n",
+    "    subplots=np.ndarray.flatten(subplots)\n",
+    "    for subplot, image in zip(subplots, digits):\n",
+    "        show_digit(image, subplot=subplot)\n",
+    "    for subplot in subplots[len(digits):]:\n",
+    "        subplot.axis('off')\n",
+    "\n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# Delete the endpoint\n",
```

Comment:
making a short text cell to explain deleting the endpoint instead of this comment might be more effective, but YMMV

Reply:
Good idea, done.